### PR TITLE
Named Entities Identification

### PR DESCRIPTION
**Pull request by Rawank**
Changed the order of the pipelines' modules:
Determining named entities after combining phrases.

### CODE REVIEW

File: qa.hawk/src/main/java/org/aksw/hawk/nlp/SentenceToSequence.java [hunk 1]
```diff
@@ -50,60 +55,263 @@ public static void runPhraseCombination(final HAWKQuestion q, final List<String>
...
 			}
 			// split "of the" or "of all" or "against" via pos_i=IN and
 			// pos_i+1=DT
 			else if (!subsequence.isEmpty() && (null != pos) && ((tcounter + 1) < question.size()) && (null != nextPos) && pos.matches("IN") && !token.matches("of")
 			        && nextPos.matches("(W)?DT|NNP(S)?")) {
-				if (subsequence.size() > 1) {
-					transformTree(subsequence, q, tokenOffset);
-				}
-				subsequence = Lists.newArrayList();
+				           namedEntityOrCombinedPhrase = checkNamedEntites(tempSubsequance2,q);
```
Comment (agarant) [inline qa.hawk/src/main/java/org/aksw/hawk/nlp/SentenceToSequence.java]: I think that this block of code is exactly the same as the one below, could you extract a function?

File: qa.hawk/src/main/java/org/aksw/hawk/nlp/SentenceToSequence.java [hunk 2]
```diff
@@ -50,60 +55,263 @@ public static void runPhraseCombination(final HAWKQuestion q, final List<String>
...
+								    {
+									subsequanceNE.add(tempSubsequance2); 
+							                transformTree(subsequanceNE, q, tokenOffset);
+								    }
+									
+							    }
+                                                 else if(isNamedEntity.contains("3")) {}
+					   }
+				tempSubsequance2 = new String();
+				subsequance = Lists.newArrayList();
```
Comment (agarant) [inline qa.hawk/src/main/java/org/aksw/hawk/nlp/SentenceToSequence.java]: subsequance is never defined? Did you mean subsequence?

File: qa.hawk/src/main/java/org/aksw/hawk/nlp/SentenceToSequence.java [hunk 3]
```diff
@@ -20,6 +20,10 @@
 	public static void runPhraseCombination(final HAWKQuestion q, final List<String> tokens, final Map<String, String> label2pos) {
 		//remove quotation marks from the sentence, then send it to the combination algo
```
Comment (agarant) [inline qa.hawk/src/main/java/org/aksw/hawk/nlp/SentenceToSequence.java:26]: You changed a lot of things in the file, does the comment still describes what this function is doing?

**Review by agarant — CHANGES_REQUESTED** [review]
The overall complexity is very high. Long functions, many nesting levels, long variable lifespans, etc. As a result, I have a hard time understanding what is the goal of your changes. Also, the indentation style is not consistent which makes it even harder to read. Please take the time to clean up this code so that it is more descriptive. Try to keep your functions short (<50 loc), avoid to have more than 1 or 2 nested levels, etc. There is also a significant amount of compiler errors that need to be addressed before it can be merged.

File: qa.hawk/src/main/java/org/aksw/hawk/nlp/SentenceToSequence.java (hunk 3 above)
Comment (Rawank) [reply qa.hawk/src/main/java/org/aksw/hawk/nlp/SentenceToSequence.java:26]: yes
each comment describes what that function or command do

**Comment by Rawank** [comment]
Before change:
* Named entities can't be combined with other phrases because they are replaced by URIs in the question
* Hawk add query fragments based on the retrieved named entities and phrases from getLanguageToNamedEntites() & getLanguageToNounPhrases() 

After change:
*Named entities are identified BUT not replaced by URIs in the question
* Combination is done for all question phrases including NEs
* all combined phrases are added to getLanguageToNounPhrases() 
* Query fragments are added based on getLanguageToNounPhrases(), which contains NEs & CNNs

**Pull request merged**